projects/telemetry/inc/nmea.h

```diff
@@ -0,0 +1,65 @@
+#pragma once
+#include <stdio.h>
+#include <stdint.h>
+#include "uart_mcu.h"
+typedef enum {
+  GGA = 15, // These are the number of data fields in each message type, excluding the message id
+  GLL = 8,
+  GSA = 10,
+  GSV = 12,
+  RMC = 13,
+  VTG = 10,
+  NUM_MESSAGE_IDS = 6,
+} NMEAMessageID;
+
+typedef struct {
+  uint32_t hh; // Hours
+  uint32_t mm; // Minutes
+  uint32_t ss; // Seconds
+  uint32_t sss; // Milliseconds
+} UTCTime;
+
+typedef struct {
+  uint32_t degrees;
+  uint32_t minutes;
+  uint32_t fraction;
+} Coord; // Valid representation of longtitude or latitude
+// Info passed from the GPS chip should be dropped into this struct (more fields coming soon)
+
+typedef struct {
+  NMEAMessageID message_id;
+  UTCTime time;
+  Coord latitude;
+  uint8_t north_south; // North or South. N for North, S for South, treat as char
+  Coord longtitude;
+  uint8_t east_west; // East or West. E for East, W for West, treat as char
+  uint32_t position_fix; // 0 = invalid, (1,2,6) = valid, all else is undefined
+  uint32_t satellites_used;
+  float hdop; // Horizontal dilution of percision
+  float msl_altitude; // In meters
+  uint8_t units_1; // Indicated units of above, should be M for meters.
+  float geoid_seperation; // In meters
+  uint8_t units_2; // Indicates units of above, should be M for meters.
+  uint32_t adc; // Age of diff. corr. in seconds
+  uint32_t drs; // Diff. Ref. Station. Not sure what it is yet
+  uint8_t checksum[3]; // Should be * followed by two integers
+} GGASentence;
+
+// This struct will have the parsed sentence in it. It will have multiple sentences as they are
+// implemented. Others will just be null on return.
+// The reason for this "strange" return type is because the function doesn't know which sentence
+// will be initially passed to it,
+// so there isn't any way to overload the function to provide multiple return types. Furthermore,
+// the code that is calling the parser
+// also doesn't know which sentence it is passing, because it is not parsing it yet.
+typedef struct {
+  union{
+    GGASentence gga;
+    // GLLSentence gll; For instance, if the parse function was returning a gga sentence, then the
```

Comment:
That's not quite how `union`s work.

Basically a `union` allows you to store various data types in the same memory location. So if you have a `union`:

```c
union {
    uint64_t can_data;
    uint32_t u32[2];
    u16_t u16[4];
    uint8_t u8[8];
};
```

Setting  `can_data` to `0xFFFFFFFFFFFFFFFF` will also mean that the values of the `union` members will also change, since they are backed by the same memory location.

Reply:
I see, I'll remove this when we split the NMEA parser to parse individual sentences (soon).

Reply:
The intention here is to use this as a tagged union.

So when you have a value of type `NMEAResult`, you can switch on `message_type`, and then access the right union value. So, for example, if `message_type` was `GGA`, then you access `.gga`.

Since every result can only be parsed to a single type of sentence, it's fine that other members will be overwritten.

Reply:
Yup, this is the intended behaviour. Regardless, we will probably end up changing it if we split the nmea parser since the return types will be know in advance